### PR TITLE
AP-465: Add new FileUpload enum to differentiate between upload modes

### PR DESCRIPTION
**Pull request by mattusifer**
This is part 1 of https://blackfynn.atlassian.net/browse/AP-465: stop trying to use the `S3File` for our 2 upload "modes" (recursive vs non-recursive) and use different types to represent/derive the data that is required by these different modes.

For example, recursive uploads require a `base_path`, but this doesn't really make sense in the context of other upload modes.

### CODE REVIEW

File: src/bf/api/client/mod.rs [hunk 1]
```diff
@@ -708,8 +713,8 @@ impl Blackfynn {
     /// Generate a preview of the files to be uploaded.
     pub fn preview_upload_using_upload_service<P, Q>(
         &self,
-        organization_id: &OrganizationId,
-        dataset_id: &DatasetId,
+        organization_id: OrganizationId,
```
Comment (mattusifer) [inline src/bf/api/client/mod.rs]: I'm going to revert this change, no need to break our existing client api as part of this PR.

File: src/bf/model/upload.rs [hunk 2]
```diff
@@ -179,48 +179,205 @@ pub struct ChunkedUploadProperties {
...
+            absolute_path.is_file(),
+            format!("File is not a regular file: {:?}", absolute_path)
+        );
+
+        Ok(FileUpload::FlatDirectoryUpload {
+            absolute_path: absolute_path.to_path_buf(),
+        })
+    }
+
+    /// Returns a FileUpload object that represents a single file
```
Comment (mikeswoods) [inline src/bf/model/upload.rs]: Awesome! This is the kind of documentation that we need (as an org) everywhere.

File: src/bf/model/upload.rs [hunk 3]
```diff
@@ -179,48 +179,205 @@ pub struct ChunkedUploadProperties {
...
-/// `metadata` is the metadata of the file being uploaded
-struct NormalizedPath {
-    file_name: String,
-    destination_path: Option<Vec<String>>,
-    metadata: fs::Metadata,
-}
+/// A type representing a file to be uploaded.
+#[derive(Clone, Deserialize, Debug, Eq, Hash, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum FileUpload {
```
Comment (mikeswoods) [inline src/bf/model/upload.rs]: Much cleaner!

File: src/bf/model/upload.rs [hunk 4]
```diff
@@ -179,48 +179,205 @@ pub struct ChunkedUploadProperties {
...
+    RecursiveDirectoryUpload {
+        base_path: PathBuf,
+        relative_path: PathBuf,
```
Comment (ptravers) [inline src/bf/model/upload.rs]: Ooof that naming! Beautiful!

**Comment by ptravers** [comment]
This is awesome Matt!
